### PR TITLE
Allow Sinks to know which column they are associated with.

### PR DESCRIPTION
**Pull request by kosak**
A feature request from Deephaven Enterprise: https://github.com/deephaven/deephaven-csv/issues/37
Also a bugfix where DateTimeAsLong and TimestampAsLong were referring to the wrong "reserved" value.
Also some small comment cleanups.

### CODE REVIEW

File: src/main/java/io/deephaven/csv/parsers/Parser.java [hunk 1]
```diff
@@ -75,6 +75,10 @@ long tryParse(
             throws CsvReaderException;
 
     class GlobalContext {
+        /**
+         * The 1-based column number that the parser is working on.
+         */
+        public final int colNum;
```
Comment (devinrsmith) [inline src/main/java/io/deephaven/csv/parsers/Parser.java:81]: Better to make this private and a getter for it IMO; or be consistent and let nullChunk be public.

File: src/main/java/io/deephaven/csv/parsers/Parser.java [hunk 2]
```diff
@@ -75,6 +75,10 @@ long tryParse(
             throws CsvReaderException;
 
     class GlobalContext {
+        /**
+         * The 1-based column number that the parser is working on.
```
Comment (devinrsmith) [inline src/main/java/io/deephaven/csv/parsers/Parser.java:79]: Is there precedent in other places for CSV columns to be 1-based?

File: src/main/java/io/deephaven/csv/parsers/Parser.java (hunk 1 above)
Comment (kosak) [reply src/main/java/io/deephaven/csv/parsers/Parser.java:81]: done

File: src/main/java/io/deephaven/csv/parsers/Parser.java (hunk 2 above)
Comment (kosak) [reply src/main/java/io/deephaven/csv/parsers/Parser.java:79]: Yeah, I was rather conflicted about this choice.
On one hand, various entry points in `CsvSpecs` take one-based indices (like `putParserForIndex`). But one-based programming feels pretty weird elsewhere. I've changed this `colNum` thing to zero-based. Let me know if you like it better.

**Review by devinrsmith — APPROVED** [review]

File: src/main/java/io/deephaven/csv/parsers/Parser.java (hunk 2 above)
Comment (devinrsmith) [reply src/main/java/io/deephaven/csv/parsers/Parser.java]: I don't think it's urgent (as it's a breaking change); but https://github.com/deephaven/deephaven-csv/issues/42

File: src/test/java/io/deephaven/csv/CsvReaderTest.java [hunk 3]
```diff
@@ -1730,7 +1730,7 @@ public int read() throws IOException {
         }
 
         @Override
-        public int read(@NotNull byte[] b, int off, int len) throws IOException {
+        public int read(byte @NotNull [] b, int off, int len) {
```
Comment (devinrsmith) [inline src/test/java/io/deephaven/csv/CsvReaderTest.java:1733]: What happened here? Is this valid annotation location?

Comment (kosak) [reply src/test/java/io/deephaven/csv/CsvReaderTest.java:1733]: IDEA was warning about the original ("Primitive type members cannot be annotated") and so I deleted the annotation and asked IDEA to re-add it, and it put it as indicated.

I *assumed* this was an IDEA bug, but then I found this: https://www.jetbrains.com/help/idea/nullable-and-notnull-annotations.html#notnull

```
If @NotNull has the _TYPE_USE_ target, it’s applied to the array element type, not to the array type itself. To annotate the array type with the TYPE_USE annotation, use the byte @NotNull [] bytes syntax.
```

I also wrote a program to look into where annotations are stored in the runtime. (see below, with output).

So I *think* the change is correct, but I admit it looks surprising.

```
import org.jetbrains.annotations.NotNull;

import java.lang.annotation.*;
import java.lang.reflect.AnnotatedType;
import java.lang.reflect.Parameter;

public class Repro {
    public static void doit0(@NotNull int[] q) {
        
    }
    public static void doit1(@A int [] x) {
    }

    public static void doit2(int @A [] x) {
    }

    public static void doit3(@A int @A [] x) {
    }

    public static void main(String[] args) throws NoSuchMethodException {
        analyze("doit1");
        analyze("doit2");
        analyze("doit3");
    }

    private static void analyze(String name) throws NoSuchMethodException {
        final Parameter p = Repro.class.getMethod(name, int[].class).getParameters()[0];
        final AnnotatedType at = p.getAnnotatedType();
        final Annotation[] paramAnnotations = p.getAnnotations();
        final Annotation[] typeAnnotations = at.getAnnotations();
        System.out.printf("##### For method %s:\n", name);
        System.out.printf("type is %s\n", at.getType().getTypeName());
        System.out.printf("There are %d parameter annotations:\n", paramAnnotations.length);
        for (Annotation a : paramAnnotations) {
            System.out.println(a.annotationType().getTypeName());
        }
        System.out.printf("There are %d type annotations:\n", typeAnnotations.length);
        for (Annotation a : typeAnnotations) {
            System.out.println(a.annotationType().getTypeName());
        }
    }
}

@Retention(RetentionPolicy.RUNTIME)
@Target({ElementType.PARAMETER, ElementType.TYPE_USE})
@interface A {
    int value() default 1;
}
```

Output:
```
##### For method doit1:
type is int[]
There are 1 parameter annotations:
A
There are 0 type annotations:
##### For method doit2:
type is int[]
There are 0 parameter annotations:
There are 1 type annotations:
A
##### For method doit3:
type is int[]
There are 1 parameter annotations:
A
There are 1 type annotations:
A
```

**Pull request merged**